dot/sync/bootstrap_syncer_test.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package sync
+
+import (
+	"testing"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_bootstrapSyncer_handleWorkerResult(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		blockStateBuilder func(ctrl *gomock.Controller) BlockState
+		worker            *worker
+		wantWorkerToRetry *worker
+		err               error
+	}{
+		"nil worker.err returns nil": {
+			blockStateBuilder: func(ctrl *gomock.Controller) BlockState {
+				return NewMockBlockState(ctrl)
+			},
+			worker: &worker{},
+		},
+		"targetNumber < bestBlockHeader number returns nil": {
+			blockStateBuilder: func(ctrl *gomock.Controller) BlockState {
+				mockBlockState := NewMockBlockState(ctrl)
+				mockBlockState.EXPECT().BestBlockHeader().Return(&types.Header{Number: 2}, nil)
+				return mockBlockState
+			},
+			worker: &worker{
+				err:          &workerError{},
+				targetNumber: uintPtr(0),
+			},
+		},
+		"targetNumber > bestBlockHeader number worker errUnknownParent returns worker": {
+			blockStateBuilder: func(ctrl *gomock.Controller) BlockState {
+				mockBlockState := NewMockBlockState(ctrl)
+				mockBlockState.EXPECT().BestBlockHeader().Return(&types.Header{Number: 2}, nil)
+				mockBlockState.EXPECT().GetHighestFinalisedHeader().Return(&types.Header{Number: 1}, nil)
+				return mockBlockState
+			},
+			worker: &worker{
+				err:          &workerError{err: errUnknownParent},
+				targetNumber: uintPtr(3),
+			},
+			wantWorkerToRetry: &worker{
+				startNumber:  uintPtr(1),
+				targetNumber: uintPtr(3),
+			},
+		},
+		"targetNumber > bestBlockHeader number returns worker": {
+			blockStateBuilder: func(ctrl *gomock.Controller) BlockState {
+				mockBlockState := NewMockBlockState(ctrl)
+				mockBlockState.EXPECT().BestBlockHeader().Return(&types.Header{Number: 2}, nil)
+				return mockBlockState
+			},
+			worker: &worker{
+				err:          &workerError{},
+				targetNumber: uintPtr(3),
+			},
+			wantWorkerToRetry: &worker{
+				startNumber:  uintPtr(3),
+				targetNumber: uintPtr(3),
+			},
+		},
+	}
+	for testName, tt := range tests {
+		tt := tt
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
+			s := &bootstrapSyncer{
+				blockState: tt.blockStateBuilder(ctrl),
+			}
+			gotWorkerToRetry, err := s.handleWorkerResult(tt.worker)
+			assert.ErrorIs(t, err, tt.err)
```

Comment:
I need help with this.... when I try to modify I get error:
```
       	Error Trace:	bootstrap_syncer_test.go:110
        	Error:      	Target error should be in err chain:
        	            	expected: "cannot get header of best block"
        	            	in chain: "cannot get header of best block"
        	Test:       	Test_bootstrapSyncer_handleWorkerResult/best_block_header_error
```

Reply:
With `assert.ErrorIs`, you need to check the 'bottom-level' error that is wrapped, not its context or the full error string.

So in this case, we have `fmt.Errorf("cannot get header of best block: %w", err)` so we need to check against the `err`, not the `cannot get header of best block:` string context. If you check in `GetHeader(hash common.Hash) (header *types.Header, err error)`, such an error would be `chaindb.ErrKeyNotFound` for example.

This is useful to assert programmatic behavior since one can use `errors.Is(err, errGlobalVar)` in calling code at any level (a bit like exceptions work in other languages I guess), whereas the full error string is really just for logs/humans.